include/xtensor/xrepeat.hpp

```diff
@@ -0,0 +1,442 @@
+****************************************************************************/
+
+#ifndef XTENSOR_XREPEAT
+#define XTENSOR_XREPEAT
+
+#include <vector>
+#include <utility>
+
+#include "xaccessible.hpp"
+#include "xiterable.hpp"
+
+
+namespace xt
+{
+    template <class CT> class xrepeat;
+    template <class S> class xrepeat_stepper;
+
+    template <class CT>
+    struct xcontainer_inner_types<xrepeat<CT>>
+    {
+        using xexpression_type = std::decay_t<CT>;
+        using reference = inner_reference_t<CT>;
+        using const_reference = typename xexpression_type::const_reference;
+        using size_type = typename xexpression_type::size_type;
+        using temporary_type = typename xexpression_type::temporary_type;
+
+        static constexpr bool is_const = std::is_const<std::remove_reference_t<CT>>::value;
+
+        using extract_storage_type = xtl::mpl::eval_if_t<has_data_interface<xexpression_type>,
+            detail::expr_storage_type<xexpression_type>,
+            make_invalid_type<>>;
+        using storage_type = std::conditional_t<is_const, const extract_storage_type, extract_storage_type>;
+    };
+
+    template <class CT>
+    struct xiterable_inner_types<xrepeat<CT>>
+    {
+        using xexpression_type = std::decay_t<CT>;
+        using inner_shape_type = typename xexpression_type::inner_shape_type;
+        using stepper = xrepeat_stepper<typename xexpression_type::stepper>;
+        using const_stepper = xrepeat_stepper<typename xexpression_type::stepper>;
+    };
+
+    template <class S>
+    class xrepeat_stepper
+    {
+    public:
+        using storage_type = typename S::storage_type;
+        using subiterator_type = typename S::subiterator_type;
+        using subiterator_traits = typename S::subiterator_traits;
+        using value_type = typename subiterator_traits::value_type;
+        using reference = typename subiterator_traits::reference;
+        using pointer = typename subiterator_traits::pointer;
+        using difference_type = typename subiterator_traits::difference_type;
+        using size_type = typename storage_type::size_type;
+        using shape_type = typename storage_type::shape_type;
+        using simd_value_type = xt_simd::simd_type<value_type>;
+
+        template <class requested_type>
+        using simd_return_type = xt_simd::simd_return_type<value_type, requested_type>;
+
+        xrepeat_stepper(S&& s, const std::vector<std::ptrdiff_t>& repeats);
+
+        reference operator*() const;
+
+        void step(size_type dim, size_type n = 1);
+        void step_back(size_type dim, size_type n = 1);
+        void reset(size_type dim);
+        void reset_back(size_type dim);
+
+        void to_begin();
+        void to_end(layout_type l);
+
+        template <class T>
+        simd_return_type<T> step_simd();
+
+        void step_leading();
+
+        template <class R>
+        void store_simd(const R& vec);
+
+        struct repeated_index
```

Comment:
I thinks this can be a regular (i.e. not inner) class in the namespace `detail`. This would remove the implementation from the "declarative" part of the header.